File: ReinforcementLearning/qlearningAgent.py
from agent import Agent
from utils import Piece, Player, Counter
import random
from data import EvaluationMatrix


class QLearningAgent(Agent, EvaluationMatrix):
    def __init__(self, direction: Player, alpha=0.5, gamma=0.8, epsilon=0.5, q_value=None):
        """
        alpha    - learning rate
        gamma    - discount factor
        numTraining - number of training episodes
        """
        super().__init__(direction)
        EvaluationMatrix.__init__(self)
        self.epsilon = float(epsilon)
        self.alpha = float(alpha)
        self.discount = float(gamma)
        self.q_value = q_value
        self.last_state = None
        self.last_action = None
        self.playerSide = direction
        self.myreward = 0

    def getState(self):
        return tuple(tuple(x) for x in self.game.getGameState().getBoard())

    def getQValueBoard(self) -> Counter:
        return self.q_value

    def getQValue(self, current_board, action: tuple[tuple[int, int], tuple[int, int]]) -> float:

        return self.q_value[(current_board, action)]

    def computeActionFromQValues(self, current_board) -> tuple[tuple[int, int], tuple[int, int]]:
        q_list = [self.getQValue(current_board, action) for action in self.game.getLegalActionsBySide(self.direction)]
        max_q = max(q_list)
        max_index = []
        for i in range(len(q_list)):
            if q_list[i] == max_q:
                max_index.append(i)
        index = random.choice(max_index)
        return self.game.getLegalActionsBySide(self.direction)[index]

    def update(self, current_action):
        # get the old estimate
        lastboard = tuple(tuple(x) for x in self.last_state.getBoard())
        old_estimate = self.getQValue(lastboard, self.last_action)
        # compute max Q
        board = tuple(tuple(x) for x in self.game.getGameState().getBoard())
        # best_action=self.computeActionFromQValues(self.game)
        best_action = self.computeActionFromQValues(board)
        # max_q = self.getQValue(self.game, best_action)
        max_q = self.getQValue(board, best_action)
        opponentReward = self.getReward(self.game.getGameState(), current_action, Player.reverse(self.playerSide))
        reward = self.myreward - opponentReward
        sample = reward + self.discount * max_q
        # update the q_value
        self.q_value[(lastboard, self.last_action)] = (1 - self.alpha) * old_estimate + self.alpha * sample

    def getReward(self, state, action: tuple[tuple[int, int], tuple[int, int]], direction) -> float:
        """
            If the action has eaten one piece, return the corresponding reward;
            Else 0
        """
        score = 0
        nextstate = state.getNextState(action)
        nextstate.swapDirection()
        winner = nextstate.getWinner()
        if winner == direction:
            score = 1000000
        elif winner == Player.reverse(direction):
            score = -1000000
        else:
            myPiece = state.getSide(direction)
            for piece in myPiece:
                x, y = piece
                pieceType = state[x][y]
                score += self.pieceValue[pieceType] * self.pieceScore[pieceType][x][y]
            eatenPieceType = self.game.board[action[1][0]][action[1][1]]
            if eatenPieceType == Piece.BSoldier or eatenPieceType == Piece.RSoldier:
                score += 1
            elif eatenPieceType == Piece.BAdvisor or eatenPieceType == Piece.RAdvisor:
                score += 2
            elif eatenPieceType == Piece.BElephant or eatenPieceType == Piece.RElephant:
                score += 2
            elif eatenPieceType == Piece.BHorse or eatenPieceType == Piece.RHorse:
                score += 4
            elif eatenPieceType == Piece.BCannon or eatenPieceType == Piece.RCannon:
                score += 4.5
            elif eatenPieceType == Piece.BChariot or eatenPieceType == Piece.RChariot:
                score += 9
        return score

    def flipCoin(self,p):
        r = random.random()
        return r < p

    def getAction(self,current_board):
        legalActions = self.game.getGameState().getLegalActionsBySide(self.playerSide)
        action = None
        if self.flipCoin(self.epsilon):
            action = random.choice(legalActions)
        else: 
            action=self.computeActionFromQValues(current_board)
        return action

    def step(self) -> tuple[tuple[int, int], tuple[int, int]]:
        # get action
        board = tuple(tuple(x) for x in self.game.getGameState().getBoard())
        action = self.getAction(board)
        # save action and state
        self.last_action = action
        self.last_state = self.game.getGameState()
        # get reward
        self.myreward = self.getReward(self.last_state, self.last_action, self.playerSide)
        nextstate=self.last_state.getNextState(self.last_action)
        nextstate.swapDirection()
        winner = nextstate.getWinner()
        if winner == self.playerSide:
            old_estimate=self.q_value[(board, self.last_action)]
            self.q_value[(board, self.last_action)] = (1 - self.alpha) * old_estimate + self.alpha * self.myreward
        return action

File: ReinforcementLearning/qlearningAgent.cpp
#include "qlearningAgent.h"
#include <algorithm>
#include <random>

Action QlearningAgent::step()
{
    Board board = game->getGameState().board;
    Action action = getAction(board);
    last_action = action;
    last_state = game->getGameState();
    myreward = getReward(last_state, last_action, direction);
    GameState nextstate = last_state.getNextState(last_action);
    nextstate.swapDirection();
    Player winner = nextstate.getWinner();
    if (winner == direction)
    {
        float old_estimate = q_value[{board, last_action}];
        q_value.set({board, last_action}, (1 - myalpha) * old_estimate + myalpha * myreward);
    }
    return action;
}

Counter QlearningAgent::getQValueBoard()
{
    return q_value;
}

void QlearningAgent::update(Action action)
{
    Board lastboard = last_state.board;
    float old_estimate = getQValue(lastboard, last_action);
    Board board = game->getGameState().board;
    Action best_action = computeActionFromQValues(board);
    float max_q = getQValue(board, best_action);
    Player opponent = direction.reverse();
    float opponentReward = getReward(game->getGameState(), action, opponent);
    float reward = myreward - opponentReward;
    float sample = reward + mydiscount * max_q;
    q_value.set({lastboard, last_action}, (1 - myalpha) * old_estimate + myalpha * sample);
}

float QlearningAgent::getQValue(const Board &current_board, Action action)
{
    return q_value[{current_board, action}];
}

Action QlearningAgent::computeActionFromQValues(Board current_board)
{
    std::vector<float> q_list;
    auto all_actions = game->getGameState().getLegalActionsBySide(direction);
    for (auto &action : all_actions)
    {
        float qvalue = getQValue(current_board, action);
        q_list.push_back(qvalue);
    }
    float max_q = *max_element(q_list.begin(), q_list.end());
    std::vector<unsigned int> max_index;

    for (unsigned int i = 0; i < q_list.size(); i++)
    {
        if (q_list[i] == max_q)
        {
            max_index.push_back(i);
        }
    }
    auto len = max_index.size();
    std::random_device dev;
    std::mt19937 rng(dev());
    std::uniform_int_distribution<std::mt19937::result_type> dist(0, len);
    unsigned int index = dist(rng);
    return all_actions[index];
}

float QlearningAgent::getReward(GameState state, Action action, Player player)
{
    float score = 0;
    GameState nextstate = last_state.getNextState(last_action);
    nextstate.swapDirection();
    Player winner = nextstate.getWinner();
    if (winner == player)
    {
        score = 1000000;
    }
    else if (winner == player.reverse())
    {
        score = -1000000;
    }
    else
    {
        std::vector<Position> myPiece = state.getSide(player);
        for (Position &piece : myPiece)
        {
            auto x = piece.first;
            auto y = piece.second;
            Piece pieceType = state[piece];
            score += (pieceValue[pieceType.value()] * pieceScore[pieceType.value()][x][y]);
        }
        Piece eatenPieceType = state[action.second];
        switch (eatenPieceType.value())
        {
        case Piece::NoneType:
        {
            break;
        }
        case Piece::BSoldier:
        {
            score += 1;
            break;
        }
        case Piece::RSoldier:
        {
            score += 1;
            break;
        }
        case Piece::BAdvisor:
        {
            score += 2;
            break;
        }
        case Piece::RAdvisor:
        {
            score += 2;
            break;
        }
        case Piece::BElephant:
        {
            score += 2;
            break;
        }
        case Piece::RElephant:
        {
            score += 2;
            break;
        }
        case Piece::BHorse:
        {
            score += 4;
            break;
        }
        case Piece::RHorse:
        {
            score += 4;
            break;
        }
        case Piece::BCannon:
        {
            score += float(4.5);
            break;
        }
        case Piece::RCannon:
        {
            score += float(4.5);
            break;
        }
        case Piece::BChariot:
        {
            score += 9;
            break;
        }
        case Piece::RChariot:
        {
            score += 9;
            break;
        }
        case Piece::BGeneral:
        {
            break;
        }
        case Piece::RGeneral:
        {
            break;
        }
        }
    }
    return score;
}

bool QlearningAgent::flipcoin()
{
    std::random_device dev;
    std::mt19937 rng(dev());
    std::uniform_int_distribution<std::mt19937::result_type> dist(0, 100);
    float r = dist(rng) / float(100.0);
    return r < myepsilon;
}

Action QlearningAgent::getAction(Board board)
{
    std::vector<Action> all_actions = game->getGameState().getLegalActionsBySide(direction);
    Action action;
    if (flipcoin())
    {
        size_t len = all_actions.size();
        std::random_device dev;
        std::mt19937 rng(dev());
        std::uniform_int_distribution<std::mt19937::result_type> dist(0, len);
        action = all_actions[dist(rng)];
    }
    else
    {
        action = computeActionFromQValues(board);
    }
    return action;
}
